src/ServiceControl.Transports.AmazonSQS/QueueLengthProvider.cs

```diff
@@ -0,0 +1,132 @@
+        Task pooler;
+
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            this.store = store;
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            var endpointInputQueue = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+            var queue = QueueNameHelper.GetSqsQueueName(metadataReport.LocalAddress);
+
+            queues.AddOrUpdate(endpointInputQueue, _ => queue, (_, currentQueue) =>
+            {
+                if (currentQueue != queue)
+                {
+                    sizes.TryRemove(currentQueue, out var _);
+                }
+
+                return queue;
+            });
+
+            sizes.TryAdd(queue, 0);
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, TaggedLongValueOccurrence metricsReport)
+        {
+            //HINT: ASQ  server endpoints do not support endpoint level queue length monitoring
+        }
+
+        public Task Start()
+        {
+            stop = new CancellationTokenSource();
+
+            pooler = Task.Run(async () =>
+            {
+                using (var client = new AmazonSQSClient())
+                {
+                    var cache = new QueueUrlCache(client);
+
+                    while (!stop.Token.IsCancellationRequested)
+                    {
+                        try
+                        {
+                            await FetchQueueSizes(cache, client).ConfigureAwait(false);
+
+                            UpdateQueueLengthStore();
+
+                            await Task.Delay(QueryDelayInterval);
+                        }
+                        catch (Exception e)
+                        {
+                            Logger.Error("Error querying sql queue sizes.", e);
+                        }
+                    }
+                }
+            });
+
+            return TaskEx.Completed;
+        }
+
+        public Task Stop()
+        {
+            stop.Cancel();
+
+            return pooler;
+        }
+
+        void UpdateQueueLengthStore()
+        {
+            var nowTicks = DateTime.UtcNow.Ticks;
+
+            foreach (var tableNamePair in queues)
+            {
+                store.Store(
+                    new[]{ new RawMessage.Entry
+                    {
+                        DateTicks = nowTicks,
+                        Value = sizes.TryGetValue(tableNamePair.Value, out var size) ? size : 0
+                    }},
+                    tableNamePair.Key);
+            }
+        }
+
+        Task FetchQueueSizes(QueueUrlCache cache, IAmazonSQS client) => Task.WhenAll(sizes.Select(kvp => FetchLength(kvp.Key, client, cache)));
+
+        async Task FetchLength(string queue, IAmazonSQS client, QueueUrlCache cache)
+        {
+            try
+            {
+                var attReq = new GetQueueAttributesRequest
```

Comment:
Would it make sense to get the queue url and create the `GetQueueAttributesRequest` request (yes it can be reused) when we see it in the `Process` method or should the process be as shortlived as possible? If that is the case maybe we can tweak it a bit so that we don't need to recreate the `GetQueueAttributesRequest` over and over again given the queues should be fairly stable over time? So maybe instead of caching the URL we cache the whole request object?

Reply:
Will do